**复习：**在前面我们已经学习了Pandas基础，知道利用Pandas读取csv数据的增删查改，今天我们要学习的就是**探索性数据分析**，主要介绍如何利用Pandas进行排序、算术计算以及计算描述函数describe()的使用。

# 1 第一章：探索性数据分析

总结
1. 值排序：sort.values(by=[],ascending=[]) true是升序 false是降序
2. 行和列排序：sort_index(axis= ,ascending=) axis 1是列 0是行
3. 创建DataFrame, 可以使用np的矩阵创建，index 和columns分别是行列索引
4. DataFrame相加时同行列索引才可以相加，否则产生nan在新的行或列中
5. describe()可以查看基本统计信息

#### 开始之前，导入numpy、pandas包和数据

In [1]:
#加载所需的库
import numpy as np
import pandas as pd

In [22]:
#载入之前保存的train_chinese.csv数据，关于泰坦尼克号的任务，我们就使用这个数据
df=pd.read_csv('train_chinese.csv')
df.head()

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   
3     4     1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4     5     0     3                           Allen, Mr. William Henry   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
0    male  22.0       1       0         A/5 21171   7.2500   NaN    S  
1  female  38.0       1       0          PC 17599  71.2833   C85    C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
3  female  35.0       1       0            113803  53.1000  C123    S  
4    male  35.0       0       0            373450   8.0500   NaN    S

### 1.6 了解你的数据吗？
教材《Python for Data Analysis》第五章

#### 1.6.1 任务一：利用Pandas对示例数据进行排序，要求升序

In [6]:
# 具体请看《利用Python进行数据分析》第五章 排序和排名 部分

#自己构建一个都为数字的DataFrame数据
df1=pd.DataFrame(np.arange(1,10).reshape(3,3),index=[3,2,1],columns=['q','w','e'])
df1

q  w  e
3  1  2  3
2  4  5  6
1  7  8  9

【代码解析】

pd.DataFrame() ：创建一个DataFrame对象 

np.arange(8).reshape((2, 4)) : 生成一个二维数组（2*4）,第一列：0，1，2，3 第二列：4，5，6，7

index=['2, 1] ：DataFrame 对象的索引列

columns=['d', 'a', 'b', 'c'] ：DataFrame 对象的索引行

【问题】：大多数时候我们都是想根据列的值来排序,所以将你构建的DataFrame中的数据根据某一列，升序排列

In [16]:
#回答代码
df1.sort_values(by=['q'],ascending=True)

q  w  e
3  1  2  3
2  4  5  6
1  7  8  9

【思考】通过书本你能说出Pandas对DataFrame数据的其他排序方式吗？
df1.sort_index(axis=0) 0是行索引，1是列索引

【总结】下面将不同的排序方式做一个总结

1.让行索引升序排序

In [11]:
#代码
df1.sort_index(axis=0,ascending=True)

q  w  e
1  7  8  9
2  4  5  6
3  1  2  3

2.让列索引升序排序


In [12]:
#代码
df1.sort_index(axis=1,ascending=True)

e  q  w
3  3  1  2
2  6  4  5
1  9  7  8

3.让列索引降序排序

In [13]:
#代码
df1.sort_index(axis=1,ascending=False)

w  q  e
3  2  1  3
2  5  4  6
1  8  7  9

4.让任选两列数据同时降序排序

In [18]:
#代码
df1.sort_values(by=['q','w'],ascending=[False,False])

q  w  e
1  7  8  9
2  4  5  6
3  1  2  3

#### 1.6.2 任务二：对泰坦尼克号数据（trian.csv）按票价和年龄两列进行综合排序（降序排列），从这个数据中你可以分析出什么？

In [29]:
'''
在开始我们已经导入了train_chinese.csv数据，而且前面我们也学习了导入数据过程，根据上面学习，我们直接对目标列进行排序即可
head(20) : 读取前20条数据

'''
df[['票价','年龄']].head()


票价    年龄
0   7.2500  22.0
1  71.2833  38.0
2   7.9250  26.0
3  53.1000  35.0
4   8.0500  35.0

In [31]:
#代码
df.sort_values(by=['票价','年龄'],ascending=[False,False]).head(20)

乘客ID  是否幸存  仓位等级                                                 姓名  \
679   680     1     1                 Cardeza, Mr. Thomas Drake Martinez   
258   259     1     1                                   Ward, Miss. Anna   
737   738     1     1                             Lesurer, Mr. Gustave J   
438   439     0     1                                  Fortune, Mr. Mark   
341   342     1     1                     Fortune, Miss. Alice Elizabeth   
88     89     1     1                         Fortune, Miss. Mabel Helen   
27     28     0     1                     Fortune, Mr. Charles Alexander   
742   743     1     1              Ryerson, Miss. Susan Parker "Suzette"   
311   312     1     1                         Ryerson, Miss. Emily Borie   
299   300     1     1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
118   119     0     1                           Baxter, Mr. Quigg Edmond   
380   381     1     1                              Bidois, Miss. Rosalie   
716   717     1     1                      Endres, Miss. Caroline Louise   
700   701     1     1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
557   558     0     1                                Robbins, Mr. Victor   
527   528     0     1                                 Farthing, Mr. John   
377   378     0     1                          Widener, Mr. Harry Elkins   
779   780     1     1  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...   
730   731     1     1                      Allen, Miss. Elisabeth Walton   
689   690     1     1                  Madill, Miss. Georgette Alexandra   

         性别    年龄  兄弟姐妹个数  父母子女个数      船票信息        票价               客舱 登船港口  
679    male  36.0       0       1  PC 17755  512.3292      B51 B53 B55    C  
258  female  35.0       0       0  PC 17755  512.3292              NaN    C  
737    male  35.0       0       0  PC 17755  512.3292             B101    C  
438    male  64.0       1       4     19950  263.0000      C23 C25 C27    S  
341  female  24.0       3       2     19950  263.0000      C23 C25 C27    S  
88   female  23.0       3       2     19950  263.0000      C23 C25 C27    S  
27     male  19.0       3       2     19950  263.0000      C23 C25 C27    S  
742  female  21.0       2       2  PC 17608  262.3750  B57 B59 B63 B66    C  
311  female  18.0       2       2  PC 17608  262.3750  B57 B59 B63 B66    C  
299  female  50.0       0       1  PC 17558  247.5208          B58 B60    C  
118    male  24.0       0       1  PC 17558  247.5208          B58 B60    C  
380  female  42.0       0       0  PC 17757  227.5250              NaN    C  
716  female  38.0       0       0  PC 17757  227.5250              C45    C  
700  female  18.0       1       0  PC 17757  227.5250          C62 C64    C  
557    male   NaN       0       0  PC 17757  227.5250              NaN    C  
527    male   NaN       0       0  PC 17483  221.7792              C95    S  
377    male  27.0       0       2    113503  211.5000              C82    C  
779  female  43.0       0       1     24160  211.3375               B3    S  
730  female  29.0       0       0     24160  211.3375               B5    S  
689  female  15.0       0       1     24160  211.3375               B5    S

【思考】排序后，如果我们仅仅关注年龄和票价两列。根据常识我知道发现票价越高的应该客舱越好，所以我们会明显看出，票价前20的乘客中存活的有14人，这是相当高的一个比例，那么我们后面是不是可以进一步分析一下票价和存活之间的关系，年龄和存活之间的关系呢？当你开始发现数据之间的关系了，数据分析就开始了。

当然，这只是我的想法，你还可以有更多想法，欢迎写在你的学习笔记中。

**多做几个数据的排序**

In [35]:
#代码
df.sort_values(by=['票价','仓位等级'],ascending=[False,False]).tail(20)

乘客ID  是否幸存  仓位等级                                姓名    性别    年龄  兄弟姐妹个数  \
818   819     0     3  Holm, Mr. John Fredrik Alexander  male  43.0       0   
843   844     0     3        Lemberopolous, Mr. Peter L  male  34.5       0   
326   327     0     3         Nysveen, Mr. Johan Hansen  male  61.0       0   
872   873     0     1          Carlsson, Mr. Frans Olof  male  33.0       0   
378   379     0     3               Betros, Mr. Tannous  male  20.0       0   
179   180     0     3               Leonard, Mr. Lionel  male  36.0       0   
271   272     1     3      Tornquist, Mr. William Henry  male  25.0       0   
302   303     0     3   Johnson, Mr. William Cahoone Jr  male  19.0       0   
597   598     0     3               Johnson, Mr. Alfred  male  49.0       0   
277   278     0     2       Parkes, Mr. Francis "Frank"  male   NaN       0   
413   414     0     2    Cunningham, Mr. Alfred Fleming  male   NaN       0   
466   467     0     2             Campbell, Mr. William  male   NaN       0   
481   482     0     2  Frost, Mr. Anthony Wood "Archie"  male   NaN       0   
674   675     0     2        Watson, Mr. Ennis Hastings  male   NaN       0   
732   733     0     2              Knight, Mr. Robert J  male   NaN       0   
263   264     0     1             Harrison, Mr. William  male  40.0       0   
633   634     0     1     Parr, Mr. William Henry Marsh  male   NaN       0   
806   807     0     1            Andrews, Mr. Thomas Jr  male  39.0       0   
815   816     0     1                  Fry, Mr. Richard  male   NaN       0   
822   823     0     1   Reuchlin, Jonkheer. John George  male  38.0       0   

     父母子女个数    船票信息      票价           客舱 登船港口  
818       0  C 7075  6.4500          NaN    S  
843       0    2683  6.4375          NaN    C  
326       0  345364  6.2375          NaN    S  
872       0     695  5.0000  B51 B53 B55    S  
378       0    2648  4.0125          NaN    C  
179       0    LINE  0.0000          NaN    S  
271       0    LINE  0.0000          NaN    S  
302       0    LINE  0.0000          NaN    S  
597       0    LINE  0.0000          NaN    S  
277       0  239853  0.0000          NaN    S  
413       0  239853  0.0000          NaN    S  
466       0  239853  0.0000          NaN    S  
481       0  239854  0.0000          NaN    S  
674       0  239856  0.0000          NaN    S  
732       0  239855  0.0000          NaN    S  
263       0  112059  0.0000          B94    S  
633       0  112052  0.0000          NaN    S  
806       0  112050  0.0000          A36    S  
815       0  112058  0.0000         B102    S  
822       0   19972  0.0000          NaN    S

In [36]:
df.sort_values(by=['票价','仓位等级'],ascending=[False,False]).head(20)

乘客ID  是否幸存  仓位等级                                                 姓名  \
258   259     1     1                                   Ward, Miss. Anna   
679   680     1     1                 Cardeza, Mr. Thomas Drake Martinez   
737   738     1     1                             Lesurer, Mr. Gustave J   
27     28     0     1                     Fortune, Mr. Charles Alexander   
88     89     1     1                         Fortune, Miss. Mabel Helen   
341   342     1     1                     Fortune, Miss. Alice Elizabeth   
438   439     0     1                                  Fortune, Mr. Mark   
311   312     1     1                         Ryerson, Miss. Emily Borie   
742   743     1     1              Ryerson, Miss. Susan Parker "Suzette"   
118   119     0     1                           Baxter, Mr. Quigg Edmond   
299   300     1     1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
380   381     1     1                              Bidois, Miss. Rosalie   
557   558     0     1                                Robbins, Mr. Victor   
700   701     1     1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
716   717     1     1                      Endres, Miss. Caroline Louise   
527   528     0     1                                 Farthing, Mr. John   
377   378     0     1                          Widener, Mr. Harry Elkins   
689   690     1     1                  Madill, Miss. Georgette Alexandra   
730   731     1     1                      Allen, Miss. Elisabeth Walton   
779   780     1     1  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...   

         性别    年龄  兄弟姐妹个数  父母子女个数      船票信息        票价               客舱 登船港口  
258  female  35.0       0       0  PC 17755  512.3292              NaN    C  
679    male  36.0       0       1  PC 17755  512.3292      B51 B53 B55    C  
737    male  35.0       0       0  PC 17755  512.3292             B101    C  
27     male  19.0       3       2     19950  263.0000      C23 C25 C27    S  
88   female  23.0       3       2     19950  263.0000      C23 C25 C27    S  
341  female  24.0       3       2     19950  263.0000      C23 C25 C27    S  
438    male  64.0       1       4     19950  263.0000      C23 C25 C27    S  
311  female  18.0       2       2  PC 17608  262.3750  B57 B59 B63 B66    C  
742  female  21.0       2       2  PC 17608  262.3750  B57 B59 B63 B66    C  
118    male  24.0       0       1  PC 17558  247.5208          B58 B60    C  
299  female  50.0       0       1  PC 17558  247.5208          B58 B60    C  
380  female  42.0       0       0  PC 17757  227.5250              NaN    C  
557    male   NaN       0       0  PC 17757  227.5250              NaN    C  
700  female  18.0       1       0  PC 17757  227.5250          C62 C64    C  
716  female  38.0       0       0  PC 17757  227.5250              C45    C  
527    male   NaN       0       0  PC 17483  221.7792              C95    S  
377    male  27.0       0       2    113503  211.5000              C82    C  
689  female  15.0       0       1     24160  211.3375               B5    S  
730  female  29.0       0       0     24160  211.3375               B5    S  
779  female  43.0       0       1     24160  211.3375               B3    S

In [4]:
#写下你的思考
后20的票价，仓位等级1的占比明显变少，票价越高的仓位等级越好，票价越低的幸存为0的越多
女性的票价会更高，男性的票价越低


#### 1.6.3 任务三：利用Pandas进行算术计算，计算两个DataFrame数据相加结果

In [10]:
# 具体请看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分

#自己构建两个都为数字的DataFrame数据

"""
我们举了一个例子：
frame1_a = pd.DataFrame(np.arange(9.).reshape(3, 3),
                     columns=['a', 'b', 'c'],
                     index=['one', 'two', 'three'])
frame1_b = pd.DataFrame(np.arange(12.).reshape(4, 3),
                     columns=['a', 'e', 'c'],
                     index=['first', 'one', 'two', 'second'])
frame1_a
"""

In [38]:
#代码
df1=pd.DataFrame(np.arange(1,10).reshape(3,3),index=['0','1','2'],columns=['q','w','e'])
df1

q  w  e
0  1  2  3
1  4  5  6
2  7  8  9

In [40]:
df2=pd.DataFrame(np.arange(1,17).reshape(4,4),index=['1','2','4','5'],columns=['q','w','r','t'])
df2

q   w   r   t
1   1   2   3   4
2   5   6   7   8
4   9  10  11  12
5  13  14  15  16

将frame_a和frame_b进行相加


In [41]:
#代码
df1+df2

e     q   r   t     w
0 NaN   NaN NaN NaN   NaN
1 NaN   5.0 NaN NaN   7.0
2 NaN  12.0 NaN NaN  14.0
4 NaN   NaN NaN NaN   NaN
5 NaN   NaN NaN NaN   NaN

【提醒】两个DataFrame相加后，会返回一个新的DataFrame，对应的行和列的值会相加，没有对应的会变成空值NaN。<br>
当然，DataFrame还有很多算术运算，如减法，除法等，有兴趣的同学可以看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分，多在网络上查找相关学习资料。

#### 1.6.4 任务四：通过泰坦尼克号数据如何计算出在船上最大的家族有多少人？

In [42]:
'''
还是用之前导入的chinese_train.csv如果我们想看看在船上，最大的家族有多少人（‘兄弟姐妹个数’+‘父母子女个数’），我们该怎么做呢？
'''
df=pd.read_csv('train_chinese.csv')
df.head()

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   
3     4     1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4     5     0     3                           Allen, Mr. William Henry   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
0    male  22.0       1       0         A/5 21171   7.2500   NaN    S  
1  female  38.0       1       0          PC 17599  71.2833   C85    C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
3  female  35.0       1       0            113803  53.1000  C123    S  
4    male  35.0       0       0            373450   8.0500   NaN    S

In [46]:
#代码
dd=df['兄弟姐妹个数']+df['父母子女个数']
dd.max()

10

【提醒】我们只需找出”兄弟姐妹个数“和”父母子女个数“之和最大的数，当然你还可以想出很多方法和思考角度，欢迎你来说出你的看法。

**多做几个数据的相加，看看你能分析出什么？**

In [4]:
#代码

In [5]:
#写下你的其他分析







#### 1.6.5 任务五：学会使用Pandas describe()函数查看数据基本统计信息

In [13]:
#(1) 关键知识点示例做一遍（简单数据）
# 具体请看《利用Python进行数据分析》第五章 汇总和计算描述统计 部分

#自己构建一个有数字有空值的DataFrame数据


"""
我们举了一个例子：
frame2 = pd.DataFrame([[1.4, np.nan], 
                       [7.1, -4.5],
                       [np.nan, np.nan], 
                       [0.75, -1.3]
                      ], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])
frame2

"""

In [48]:
#代码
df3=pd.DataFrame([[1,np.nan,2],
                [2,3.1,np.nan],
                [np.nan,np.nan,3]],index=[1,2,3],columns=['a','b','c'])
df3

a    b    c
1  1.0  NaN  2.0
2  2.0  3.1  NaN
3  NaN  NaN  3.0

调用 describe 函数，观察frame2的数据基本信息

In [54]:
#代码
df3.describe()

a    b         c
count  2.000000  1.0  2.000000
mean   1.500000  3.1  2.500000
std    0.707107  NaN  0.707107
min    1.000000  3.1  2.000000
25%    1.250000  3.1  2.250000
50%    1.500000  3.1  2.500000
75%    1.750000  3.1  2.750000
max    2.000000  3.1  3.000000

#### 1.6.6 任务六：分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据，你能发现什么？

In [ ]:
'''
看看泰坦尼克号数据集中 票价 这列数据的基本统计数据
'''

In [55]:
#代码
df.describe()

乘客ID        是否幸存        仓位等级          年龄      兄弟姐妹个数      父母子女个数  \
count  891.000000  891.000000  891.000000  714.000000  891.000000  891.000000   
mean   446.000000    0.383838    2.308642   29.699118    0.523008    0.381594   
std    257.353842    0.486592    0.836071   14.526497    1.102743    0.806057   
min      1.000000    0.000000    1.000000    0.420000    0.000000    0.000000   
25%    223.500000    0.000000    2.000000   20.125000    0.000000    0.000000   
50%    446.000000    0.000000    3.000000   28.000000    0.000000    0.000000   
75%    668.500000    1.000000    3.000000   38.000000    1.000000    0.000000   
max    891.000000    1.000000    3.000000   80.000000    8.000000    6.000000   

               票价  
count  891.000000  
mean    32.204208  
std     49.693429  
min      0.000000  
25%      7.910400  
50%     14.454200  
75%     31.000000  
max    512.329200

【思考】从上面数据我们可以看出，试试在下面写出你的看法。然后看看我们给出的答案。

当然，答案只是我的想法，你还可以有更多想法，欢迎写在你的学习笔记中。

**多做几个组数据的统计，看看你能分析出什么？**

891名乘客,其中38.38%的乘客幸存，超过一半的人住在3等舱，平均一张票32元



【思考】有更多想法，欢迎写在你的学习笔记中。

【总结】本节中我们通过Pandas的一些内置函数对数据进行了初步统计查看，这个过程最重要的不是大家得掌握这些函数，而是看懂从这些函数出来的数据，构建自己的数据分析思维，这也是第一章最重要的点，希望大家学完第一章能对数据有个基本认识，了解自己在做什么，为什么这么做，后面的章节我们将开始对数据进行清洗，进一步分析。